src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -206,5 +292,119 @@ public Task UnloadAsync()
 
             return Task.CompletedTask;
         }
+
+        // NuGet calls this method to wait project to nominate restoring.
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project nominate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to nominate a restore (DT build failed)
+        public Task WhenNominated(CancellationToken cancellationToken)
+        {
+            lock (SyncObject)
+            {
+                if (!CheckIfHasPendingNomination() || cancellationToken.IsCancellationRequested)
+                {
+                    return Task.CompletedTask;
+                }
+
+                if (_whenNominatedTask is null)
+                {
+                    _whenNominatedTask = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                    _ = SourceBlock.Completion.ContinueWith(t =>
+                    {
+                        lock (SyncObject)
+                        {
+                            if (t.IsFaulted)
+                            {
+                                _whenNominatedTask.SetException(t.Exception);
+                            }
+                            else
+                            {
+                                _whenNominatedTask.TrySetCanceled();
+                            }
+                        }
+                    }, cancellationToken,
+                       TaskContinuationOptions.OnlyOnFaulted,
+                       TaskScheduler.Default);
+                }
+            }
+
+            return _whenNominatedTask.Task.WithCancellation(cancellationToken);
+        }
+
+        // True means the project system plans to call NominateProjectAsync in the future.
+        bool IVsProjectRestoreInfoSource.HasPendingNomination
+        {
+            get => CheckIfHasPendingNomination();
+        }
+
+        private bool CheckIfHasPendingNomination()
+        {
+            lock (SyncObject)
+            {
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+                if (SourceBlock.Completion.IsFaulted)
+                {
+                    return false;
+                }
+
+                // Avoid possible deadlock.
+                // Because RestoreCoreAsync() is called inside a dataflow block it will not be called with new data
+                // until the old task finishes. So, if the project gets nominating restore, it will not get updated data.
+                if (IsNugetRestoreFinished())
```

Comment:
This is incorrect, and happens to be the reversed logic as we expected.  We want to return false, if we are in the middle of a nuget restoring work.